### ISSUE TITLE
Extended diagnostics in compiler

### ISSUE THREAD

**Issue by dgp1130** [angular/angular#42966]
### Which @angular/* package(s) are relevant/releated to the feature request?

compiler-cli

### Description

The compiler should have some capability for "extended template diagnostics", meaning diagnostics applied after type checking. Currently, the compiler doesn't really have any warnings, and only ever fails for fatal issues that directly prevent compilation (with some minor exceptions). These extended diagnostics will allow warnings to be easily authored to check for smaller mistakes like getting two-way binding syntax backwards `([foo])="bar"` or having extraneous operators like using `foo ?? 'bar'` when `foo` is not nullable.

For scoping, we just need a private flag in the compiler for now. This will be enough for the CLI and language service to plug in and surface these diagnostics.

### Proposed solution

A new private flag in the compiler which enables "extended template diagnostic" checks that give warning/information diagnostics about user templates which aren't strictly fatal errors. This issue is mostly about building the required infrastructure, but we should include one or two simple checks to validate the system.

### Alternatives considered

N/A

**Comment by itea-dev** [comment]
@eneajaho @ahnpnl

A suggestion for you... Use thumbs up (👍) or voting.

**Comment by dgp1130** [comment]
Most of the implementation here is done (thanks @danieltre23!). We're current working through some of the surrounding integrations that need to be done. The three main tasks left before we can do an initial MVP release are:

* [ ] Determining what kind of configuration we want to support and implementing it.
* [ ] Determining how the CLI should integrate with extended template checks.
* [ ] Decide what extended checks we want to include to justify the public launch. We currently check for invalid banana in box and a useless nullish coalescing, is that enough for the initial release or should we include more?

Some initial ideas of additional checks we could add:

* [No negated async pipes](http://codelyzer.com/rules/template-no-negated-async/)
* Verify that all UI text nodes have an `i18n` attribute.
* a11y attribute validator
* Non-nullable optional chaining
    * `foo?.bar` is useless if `foo` is not nullable.
* Catching `foo | date: "format" ?? 'test'`
    * The `??` applies to `"format"`, which isn't what the user wanted.
    * Does the existing nullish coalescing check already catch this case? My intuition is yes?
* Other [Codelyzer](http://codelyzer.com/) checks.

We'll need to decide how much to do for the initial release and what we want to defer for later.

**Comment by dgp1130** [comment]
Some more possible ideas: https://github.com/angular/angular/issues/42690

**Comment by dgp1130** [comment]
Just discussed the configuration support we'll need for MVP. Our current plan is to support:

* Enabling / disabling all extended template checks via a single option.
* Suppressing individual checks by name (error codes already have "names", we just need to plumb this through and expose them publicly).
* Promoting warnings to errors by name.
* Promoting all warnings to errors via a single option.
* Document all this and the specific checks for each.

This will be managed in the `tsconfig.json` so editors, language service, and the like will see and use these options. The Angular CLI already supports compiling with a specific `tsconfig.json` in the browser builder, so users could create multiple targets with different configurations if they want to be more specific (such as warnings in local dev but errors on CI).

This should be enough for an MVP of the feature that can be officially launched. I also have #43582 out to make `ngc-wrapped` pass the build if only warnings are emitted, as well as #43572 to fix a minor bug in the nullish coalescing not nullable check. We'll also need to confirm that the CLI doesn't fail when given only warnings, but that should be an easy fix if not.

Tracking google3 launch of this feature in http://b/199443160.

**Comment by dgp1130** [comment]
https://github.com/angular/angular/pull/44391 has the main implementation of the configuration. Once that lands a simple flag flip should be sufficient to enable the feature.

The one other task left to do is make some documentation on the implemented diagnostics. Long term, we'd like to make something similar to angular.io/errors. For now, we'll start with just a single page that talks about these diagnostics and lists out the two we're starting with (`invalidBananaInBox` and `nullishCoalescingNotNullable`), how they work, why are bad practice, and what users can do about them.

**Comment by JoostK** [comment]
This is now available in 13.2, for anyone interested [here's the docs](https://angular.io/extended-diagnostics).

**Comment by dgp1130** [comment]
I opened a few issues for some of the ideas we've had about future extended diagnostics check. Please upvote if you'd like to see one of these in a future version of Angular!

* #42690
* #44867
* #44868
* #44869
* #44870
* #44871
* #44872

**Comment by DaSchTour** [comment]
I wonder what is the difference between diagnostics checks and @angular-eslint/eslint-plugin-template? I think it would be better to have one solution and not create different in worst case incompatible solutions.
Linter would also have the benefit of beeing visible while without angular runing and also having fixers.

**Comment by dgp1130** [comment]
@DaSchTour, see https://github.com/angular/angular/pull/44712#issuecomment-1012666051 for a bit more background.

Extended diagnostics are intended to work with `angular-eslint` and provide a hook for surfacing critical diagnostics to the user earlier to save debugging and implementation time. Linters offer a much broader set of checks than extended diagnostics are intended to support.

To address your two specific comments, extended diagnostics are visible in the Angular language service and should appear in your IDE. I also just opened an issue about supporting fixers (https://github.com/angular/angular/issues/44941), which is something we experimented with a while back and are hoping to implement in the future.
